### PR TITLE
RSDK-6483 Workspace-ify micro-rdk

### PR DESCRIPTION
**Pull request by acmorrow**
This is a draft because I'd like to share what I have before investing more time. I've made it a draft because right now it doesn't quite work due to some issues (I believe) around `sdkconfig.defaults`, but I expect those are solvable. There are probably other rough edges to sand down as well before this is something I'd want to merge.

If we do like this direction, we can either stop here, or continue with splitting up `micro-rdk` into a few crates. I'm envisioning something like:
- `micro-rdk-api`: The code generated by `buf` from the `api` repo
- `micro-rdk-core`: Common type definitions and support code. Modules would build against this.
- `micro-rdk-resources` or `micro-rdk-builtins`: Implementations for builtins like the ultrasonic sensor, ina*, etc.
- `micro-rdk-server`: The binary we make available for download.

### CODE REVIEW

File: Cargo.toml
Comment (npmenard) [inline Cargo.toml:20]: I think it make sense for now to keep the patches here for now since examples & micro-rdk needs them when building thier respective targets.

File: Cargo.toml
Comment (npmenard) [inline Cargo.toml:17]: This will be shared with all crates in the worksapce? Can it be overwritten if we wanted to? (ex may micro-rdk-installer will need a more recent version of rust than the one used by micro-rdk?)

File: Cargo.toml
Comment (npmenard) [inline Cargo.toml:5]: nit rename examples to micro-rdk-examples

**Review by npmenard — COMMENTED** [review]
I like this approach seems more readable and easier to navigate for a newcomer.

I don't think we need to split micro-rdk further at least for this PR, do you think it will be much work if we do it later?

Also we need to keep the Cargo.lock at least for examples & micro-rdk-installer so build can be reproduce easily can we keep them?

**Comment by acmorrow** [comment]
> I like this approach seems more readable and easier to navigate for a newcomer.

Great!

> 
> I don't think we need to split micro-rdk further at least for this PR, do you think it will be much work if we do it later?
> 

No, I think we should get the work from this review across the line and make sure we are happy living in a workspace'd world. Then we can plan out the next step of splitting up the `micro-rdk` crate as follow-on work.


> Also we need to keep the Cargo.lock at least for examples & micro-rdk-installer so build can be reproduce easily can we keep them?

Within a workspace there is only one top-level `Cargo.lock`. It is a bit of a puzzle to me, actually, how this is supposed to work for a workspace that has some libraries (lockfile should not be commited) and some programs (lockfile should be committed). I'll look into that and follow up.

**Review by gvaradarajan — COMMENTED** [review]
I agree with this layout conceptually (provided the macros can still be imported properly for modules and the installer is able to build) and I'm eager to have the crate splitting as a follow-up. The single Cargo.lock file does give me pause, but maybe it's ok if we don't have a lock file at all?

File: examples/Cargo.toml
Comment (npmenard) [reply examples/Cargo.toml:74]: done

File: micro-rdk/src/esp32/i2c.rs
Comment (npmenard) [reply micro-rdk/src/esp32/i2c.rs]: Nope @gvaradarajan what is the right syntax?

**Comment by npmenard** [comment]
> @npmenard - This LGTM. One or two small comments.
> 
> How do you want to proceed? If you like, I can take on getting this rebased (there are conflicts) and merged.
> 
> I think you will need to hit the approve button though, since I can't approve my own PR...

I just merged the diff still need to check that changes to the CI are correct. You can merge without approval :)

**Comment by acmorrow** [comment]
> > @npmenard - This LGTM. One or two small comments.
> > How do you want to proceed? If you like, I can take on getting this rebased (there are conflicts) and merged.
> > I think you will need to hit the approve button though, since I can't approve my own PR...
> 
> I just merged the diff still need to check that changes to the CI are correct. You can merge without approval :)

Great, I'll merge if/when the CI checks pass.

**Comment by acmorrow** [comment]
@npmenard - Looks like the clippy run didn't quite work: https://github.com/viamrobotics/micro-rdk/actions/runs/8158406229/job/22300277344?pr=149

**Review by npmenard — APPROVED** [review]
LGTM

**Pull request merged**
